packages/expo-device/ios/EXDevice/EXDevice.m

```diff
@@ -0,0 +1,252 @@
+        }
+      }
+      temp_addr = temp_addr->ifa_next;
+    }
+  }
+  // Free memory
+  freeifaddrs(interfaces);
+  resolve(address);
+}
+
+UM_EXPORT_METHOD_AS(isPinOrFingerprintSetAsync, isPinOrFingerprintSetAsyncWithResolver:(UMPromiseResolveBlock)resolve rejecter:(UMPromiseRejectBlock)reject)
+{
+#if TARGET_OS_TV
+  BOOL isPinOrFingerprintSet = false;
+#else
+  LAContext *context = [[LAContext alloc] init];
+  BOOL isPinOrFingerprintSet = ([context canEvaluatePolicy:LAPolicyDeviceOwnerAuthentication error:nil]);
+#endif
+  resolve(@[[NSNumber numberWithBool:isPinOrFingerprintSet]]);
+}
+
+UM_EXPORT_METHOD_AS(getFreeDiskStorageAsync, getFreeDiskStorageAsyncWithResolver:(UMPromiseResolveBlock)resolve rejector:(UMPromiseRejectBlock)reject)
+{
+  NSString *returnString = [NSString stringWithFormat:@"%llu", self.freeDiskStorage];
+  resolve(returnString);
+}
+
+- (NSString *)deviceId
+{
+  struct utsname systemInfo;
+  
+  uname(&systemInfo);
+  
+  NSString* deviceId = [NSString stringWithCString:systemInfo.machine
+                                          encoding:NSUTF8StringEncoding];
+  
+  if ([deviceId isEqualToString:@"i386"] || [deviceId isEqualToString:@"x86_64"] ) {
+    deviceId = [NSString stringWithFormat:@"%s", getenv("SIMULATOR_MODEL_IDENTIFIER")];
+  }
+  
+  return deviceId;
+}
+
+- (NSString *)carrier
+{
+#if (TARGET_OS_TV)
+  return nil;
+#else
+  CTTelephonyNetworkInfo *netinfo = [[CTTelephonyNetworkInfo alloc] init];
+  CTCarrier *carrier = [netinfo subscriberCellularProvider];
+  return carrier.carrierName;
+#endif
+}
+
+/*device types*/
+typedef NS_ENUM(NSInteger, DeviceType) {
+  DeviceTypeHandset,
+  DeviceTypeTablet,
+  DeviceTypeTv,
+  DeviceTypeUnknown
+};
+
+#define DeviceTypeValues [NSArray arrayWithObjects: @"Handset", @"Tablet", @"Tv", @"Unknown", nil]
+
+- (DeviceType)getDeviceType
+{
+  switch ([[UIDevice currentDevice] userInterfaceIdiom]) {
+    case UIUserInterfaceIdiomPhone: return DeviceTypeHandset;
+    case UIUserInterfaceIdiomPad: return DeviceTypeTablet;
+    case UIUserInterfaceIdiomTV: return DeviceTypeTv;
+    default: return DeviceTypeUnknown;
+  }
+}
+
+- (bool) isTablet
+{
+  return [self getDeviceType] == DeviceTypeTablet;
+}
+
+- (NSString *)getCPUType {
+  /* https://stackoverflow.com/questions/19859388/how-can-i-get-the-ios-device-cpu-architecture-in-runtime */
+  const NXArchInfo *info = NXGetLocalArchInfo();
+  NSString *typeOfCpu = [NSString stringWithUTF8String:info->description];
+  return typeOfCpu;
+}
+
+
+- (unsigned long long) totalMemory {
+  return [NSProcessInfo processInfo].physicalMemory;
+}
+
+- (NSDictionary *) getStorageDictionary {
+  NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
+  return [[NSFileManager defaultManager] attributesOfFileSystemForPath:[paths lastObject] error: nil];
+}
+
+- (uint64_t) totalDiskCapacity {
```

Comment:
- I updated this to return `NSNumber *` where appropriate.

>  I think (calling methods on nil is safe, idk what about accessing properties like this, you'd need to check)
- I tried to hardcode `- (NSDictionary *)documentFileSystemAttributes ` to return `nil` to test this theory, but at runtime, it made all the other `constantsToExport` return `undefined`. I'm not sure if I'm missing something here, but it seems like there may be an issue with accessing properties on `nil`. So to be safe I decided to just first check that `storage` is not `nil`, then return the `FileSystemSize`, otherwise return `nil` for the function.